### ISSUE TITLE
Custom network and layer freezing

### ISSUE THREAD

**Issue by vbelus** [tensorforce/tensorforce#627]
Hi,

I want to build a custom environment in which an action would be a 2d matrix (basically a b/w image), and one of the solutions I found is to use a policy based algorithm such as PPO with the policy network having layers of deconvolutions (I would probably use a U-net).

I first intended to use baselines, but I want the output of my network to match the action pixel-wise (the output at position (x,y) is used for the value at position (x,y) in the action), which I believe is not the case in the PPO2 implementation of baselines, where there is a fully-connected layer when the output of the network becomes the parameters of a probability distribution from which the action is sampled. 

Would it be possible to simply write the U-net architecture as a dictionary in your implementation, and have it working like I want, given the action space and network output shape are matching, or am I missing something ?

Also, is it possible to freeze layers of the network, and/or use a pre-trained network ?

I read through the documentation, but some of my questions will probably have an obvious answer somewhere in the repo, sorry for that !

**Comment by AlexKuhnle** [comment]
First, regarding freezing layers and using a pretrained network: The former is possible (may require the addition of a `trainable` flag to some layers, but in principle is definitely supported). The latter should also possible, but I haven't gotten around to look into this functionality properly. I don't expect it to be a big deal, and would be happy to finally get it done soon.

Now to your main question: By default, Tensorforce works similar to baselines in that case, however, you can still have 2d-shaped actions (but not "conv-deconv style"). PPO would still require distributions, DQN doesn't, but either way, it would certainly be possible to not go via an implicit dense layer as part of the distribution, but do something more like what you describe (so e.g. estimate distribution parameters via (de)convolutions and avoid flattening). If you want to implement something yourself, I can point you to the right places in the code, or I can help from the Tensorforce-side...

**Comment by AlexKuhnle** [comment]
Oops, closing was unintentional. :-)

**Comment by AlexKuhnle** [comment]
I actually looked into the distribution stuff, and realized that there is (hopefully) a cleaner solution. In the latest commit I removed the constraint for the network to produce an embedding vector and the distribution to expand from that. As can be seen in the new [convolution test](https://github.com/tensorforce/tensorforce/blob/master/test/test_layers.py#L25), as long as the network output size corresponds to the action size, it doesn't need to be flattened anymore, and in the [distributions](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/core/distributions/gaussian.py#L68) it is now handled by a basic linear layer. That may already be enough for you to get started? If you implement a deconv layer, would be great if you could contribute it... :-)

**Comment by vbelus** [comment]
From what I understand, isn't there still an embedding -> distribution process which needs flattening with that linear layer ? 
What's critical in my case is for the action to have some spatial information in a way an image has, which means I'd need the transition between the embedding tensor and the distribution to be pixel by pixel or at least to mind the spatial information in the same way a convolution does.

For the deconvolution, I implemented it in the same file as the standard convolutions and sent you a message on gitter before doing a pull request as per the guidelines.

Thanks a lot by the way, you're being very helpful :)

**Comment by AlexKuhnle** [comment]
The recent commit "generalizes" the `'linear'` layer to accept higher-rank tensors as well, in which case it doesn't apply a normal matrix multiplication, but a 1x1 convolution (so a linear transformation per "pixel"). That means that if your network uses conv-...-deconv-... and doesn't destroy the spatial structure, the distribution will keep it and just linearly transform each "pixel position" independently. Currently this only works well with a single input state, but as the test illustrates, you could still have multiple output actions (as long as they share the same spatial structure), hence the linear transformation per distribution.

**Comment by vbelus** [comment]
Alright then, this is exactly what I wanted :)
To give you some context, what I want to implement is close to something I did using stable_baselines in this paper https://arxiv.org/abs/1910.07788 (to see : the second method illustrated on page 5, where we are using a 1d convolution). Using convolutions this way has possibly interesting implications in physics. I'll get back to you after implementation with your library, in the meantime, thanks again.

**Comment by AlexKuhnle** [comment]
So the agent should ideally learn to produce just zeros, right? How exactly do you define your action space?

Comments:
- PPO samples its actions, which means it will never reliably produce only zeros. For a quick experiment you can set the `temperature` to `0.0` [here](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/agents/ppo.py#L232).
- `size=0` is like `size=1` plus getting rid of that rank.
- The network output doesn't need to be of exactly the size of your actions, so you could have e.g. `size=10` as the last layer. The internal distributions bring the output into the expected shape (if certain aspects of the shape are compatible, like `(n, n...)`.
- Conv layers with `window=1` are just like "higher-rank" dense layers, are you aware of that? So right now the input `(1,)` goes to `(4,)` and then to `(1,)`. Not too much "space" there, but well, the problem is also not particularly difficult. Anyway, what about `window=3` and `size=32` or so?

**Comment by vbelus** [comment]
Ideally, it should learn to apply the opposite function on my input, `x -> -x`. Producing only 0 would work too if it was deterministic.

My action space and state space are the same, defined as 
`dict(type='float', shape=(length_matrix, length_matrix, 1))`
Getting rid of that 1 still triggers an exception, and `size=0` in my last layer as well.

> Conv layers with window=1 are just like "higher-rank" dense layers

`window` is the size of the kernel and `size` the number of filters, right ? So you're right, it's only really a single MLP which when applied to input element at position (x, y), gives output element at position (x, y).
That's exaclty what I want though as there is no correlation of information between the different elements of my input.

**Comment by AlexKuhnle** [comment]
Okay, makes sense. But the state is just the result of adding many actions to a zero-matrix, right? Or does the environment also change the matrix?

Note that learning the identity function is actually not that easy for neural networks (see the [ResNet paper](https://arxiv.org/abs/1512.03385) for instance). There is a chance that maybe this simple toy example is not so simple for the model. Did changing the `temperature` argument help?

**Comment by vbelus** [comment]
You have it right, the state (initially a zero-matrix) is modified only by the actions. 

> Did changing the temperature argument help?

After changing the temperature the reward would get stuck at a low value, probably a local minima for the network.
> Note that learning the identity function is actually not that easy for neural networks

That may be the main reason for the poor performance then, and that would explain why my model is performing best when my network's architecture is a single 1x1 convolution, and not something more complicated. 

Also, changing execution from
`state = tanh(state + action)` to `state = tanh(state - action)` improves my model significantly - approximating `id` is probably easier than approximating `-id`.

Again, thanks for your help !

**Comment by AlexKuhnle** [comment]
No problem. It's worrying that such simple functions can be difficult to learn, but from my experience it's most useful to test with toy problems that nonetheless have some degree of "realistic" real-world-like complexity. Identity functions can apparently be a bit too trivial in that sense. Anyway, let me know about other questions that come up while playing around with that example.

**Comment by vbelus** [comment]
I suspect there's something I'm doing wrong with my convolutions. I already had a similar problem on my previous project, where I had to stop using convolutions with stable_baselines because they would just not train properly.

So I made another environment+test, completely analogous to this one except I replaced the matrix by a single float, and convolutions layers by dense layers, with the same number of neurons that my convolutions had filters. So basically, same problem, lower dimension, or so I thought.

Here are my files https://github.com/vbelus/tensorforce/tree/master/test_deconv
In `env.py` and `env-mlp.py` are my environments. They are the same classes except `self.matrix` is replaced by `self.value` in the mlp version.
I decided to normalize my network's input and output to make sure there is no problem from that so they are between 0 and 1. The reward is also between 0 and 1.

And the mlp version of my test finds an acceptable policy, the reward oscillating close to 0.9. With convolutions, I still can't get a reward higher than 0.5. I'm starting to think I'm just making a stupid mistake and not seeing it, but I can't figure it out. 

I know I'm starting to borrow a lot of your time, but can you please take a quick look at my files ? That would be a huge help.

**Comment by AlexKuhnle** [comment]
There seems to be a bracketing problem [here](https://github.com/vbelus/tensorforce/blob/master/test_deconv/env.py#L37), division before subtraction, although it won't matter if `n_step=1`. And I also don't know whether `abs(...)` works as expected when applied to a numpy array?

Have you also experimented with PPO, and/or with unconstrained actions, without `min/max_value`? Side comment: it's not necessary to have actions from `[0, 1]`, you can just configure `[-1, 1]`... ;-)

Can't see an obvious problem otherwise. What if you don't take the average, but the minimum over all the cells in the matrix? The runner averages over a range of 10 episodes by default, so that may affect things somewhat. Actually, I think it's probably better to turn this averaging off as default, will probably do so with the next update.

**Comment by AlexKuhnle** [comment]
(I can maybe have a go at it myself over the weekend, just to see whether I spot something...)

**Comment by vbelus** [comment]
> There seems to be a bracketing problem here, division before subtraction, although it won't matter if n_step=1. And I also don't know whether abs(...) works as expected when applied to a numpy array?

I'll change that bracketing problem right away, `abs` is working fine on numpy arrays but I'll use ``np.absolute`` to be extra safe.

I had unbounded actions at first and the problem was already there but I'll try again to be sure.

> What if you don't take the average, but the minimum over all the cells in the matrix?

I will try, but still, even if the runners averages overs 10 episodes, the reward is always > 0.7 in the single_value test. If the weights that the MLP learned in the single_value model were put into the convolutions, we should have a reward > 0.7 in the matrix test too. It would fluctuate less, sure, but still it wouldn't be consistently as low as 0.4.

By the way, normalization of the input gave better results (but didn't solve the problem) when I tried to implement a similar thing in stable_baselines, as convolutions in their implementation expect to see bitmap images, so values between 0 and 255. I don't see anything like that in your code but I'm just giving ideas.
Otherwise, maybe it has to do with the way convolutions learn.

> (I can maybe have a go at it myself over the weekend, just to see whether I spot something...)

That would be a huge help !

**Comment by AlexKuhnle** [comment]
Re convolutions: they don't assume any particular input, and hence don't do any implicit normalizations or so. You can normalize the input by explicitly configuring so, have a look at the preprocessing argument and the normalization layers. But I wouldn't expect this to help massively here...

**Comment by vbelus** [comment]
I have a few other questions -
- About the reward, I try to ensure the step reward is bounded so that the discounted sum of rewards at the end of an episode is between 0 and 1, does this have any importance ?
- I noticed ParallelRunner is not parallel in the sense of multiprocessing, and Jean's solution for that has some overhead because of sockets and is not very straightforward. Is a ParallelRunner with multiprocessing in consideration ?

**Comment by AlexKuhnle** [comment]
To (1): It certainly has an effect, and I'm not sure what's the best aspect to normalize for. In the case of PPO it probably makes sense to normalize for (discounted) cumulative sum, however, no need to keep it between 0 and 1, -1 and 1 is also fine (or maybe even a bit better, since 0-centered).

To (2): Yes, multiprocessing would be great to support, too. Jean's solution is for comparatively very expensive simulations, whereas multiprocessing is better for comparatively very cheap simulations. The main reason why it doesn't exist currently is that I've never really worked with these libraries and haven't yet found enough time to work through it. :-/

**Comment by vbelus** [comment]
@AlexKuhnle I added this wrapper to train in parallel with mulitprocessing in my branch https://github.com/vbelus/tensorforce/blob/master/tensorforce/environments/environment_process_wrapper.py
An example of usage with a custom tensorforce environment I have, called `SimplePendulumEnv` :
```python3
from env import SimplePendulumEnv
from tensorforce.execution import ParallelRunner
from tensorforce.environments.environment_process_wrapper import ProcessWrapper
import os

n_env = 8
list_envs = []

for index in range(n_env):
    list_envs.append(ProcessWrapper(SimplePendulumEnv()))

agent=...
agent.initialize()

# Initialize the runner
runner = ParallelRunner(agent=agent, environments=list_envs)

# Start the runner
runner.run(num_episodes=30000, sync_episodes=False)
runner.close()
```

So basically you only need to add this wrapper and it works as it was before. I have yet to benchmark it and make sure it's working as intended (looks like it is). When it's done, what about a PR ?

**Comment by AlexKuhnle** [comment]
Hey, that sounds great, a PR on that is very welcome. How are you planning to benchmark it, out of interest? There may be subtleties with how the ParallelRunner works that may influence things, not sure. Also, do you know how the ParallelRunner works via these threaded Environment functions [here](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/environments/environment.py#L191). I'm not very experienced with multiprocessing etc, so I just want to point this out.

**Comment by vbelus** [comment]
I just want to check how much faster it is compared to the multithreaded version. on increasingly complex environment (in the sense that it would take more and more time computing one step of environment). I expect the multiprocessing version to be especially efficient in cases where the environment step is the bottleneck of the computation, as it is the only computation happening on separate processes so on separate CPUs.
In my SimplePendulumEnv case, the environment is so simple the two solutions (multiprocessing and threading) should be roughly similar in terms of performance, as in my multiprocessing implementation the model can still only act for one environment at a time.

**Comment by vbelus** [comment]
@AlexKuhnle 
Back to my custom network issues, I would need to reshape my tensors in the architecture, but I don't see how I could do that with tensorforce layers without flattening.
In my specific case, I need a simple reshape from (10, 1, 128) to (10, 128).
A reshape layer is probably a layer that can be added to the repo, however is it not simpler to build my network using tensorflow/keras directly ?
I saw here https://github.com/tensorforce/tensorforce/issues/488 that it should be possible, but I don't get how to do it.

**Comment by AlexKuhnle** [comment]
There is also the possibility to act for multiple environments at each time, but that feature is currently not implemented (but hopefully soon).

For Keras layers, it's possible to use [this wrapper layer](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/core/layers/keras.py). Entire Keras networks not yet, would need to look into that... For other kind of reshaping, I think the [pooling layer](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/core/layers/pooling.py#L24) will do, but currently it collapses all dimensions, so will need to improve its interface a bit.

**Comment by AlexKuhnle** [comment]
I think by now all the features mentioned here are implemented. Re-open if not... :-)
